stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/utils/Linkify.kt

```diff
@@ -0,0 +1,214 @@
+     *
+     * @param textView TextView whose text is to be marked-up with links.
+     */
+    public fun addLinks(textView: TextView) {
+        val t: CharSequence = textView.text
+
+        if (t is Spannable) {
+            if (addLinks(t)) {
+                addLinkMovementMethod(textView)
+            }
+        } else {
+            val s = SpannableString.valueOf(t)
+            if (addLinks(s)) {
+                addLinkMovementMethod(textView)
+                textView.text = s
+            }
+        }
+    }
+
+    /**
+     * Scans the provided spannable text and turns all occurrences
+     * of the link types into clickable links (Currently only support web urls).
+     *
+     * @param text Spannable whose text is to be marked-up with links.
+     * @return True if at least one link is found and applied.
+     */
+    @SuppressLint("RestrictedApi")
+    private fun addLinks(text: Spannable): Boolean {
+        val links = mutableListOf<LinkSpec>()
+        gatherLinks(
+            links, text, PatternsCompat.AUTOLINK_WEB_URL, arrayOf("http://", "https://", "rtsp://"),
+            Linkify.sUrlMatchFilter, null
+        )
+
+        pruneOverlaps(links, text)
+
+        if (links.isEmpty()) return false
+
+        links.forEach { link ->
+            if (link.markwonAddedSpan == null) {
+                applyLink(link.url!!, link.start, link.end, text)
+            }
+        }
+
+        return true
+    }
+
+    private fun addLinkMovementMethod(t: TextView) {
+        val m = t.movementMethod
+        if (m !is LinkMovementMethod) {
+            if (t.linksClickable) {
+                t.movementMethod = LinkMovementMethod.getInstance()
+            }
+        }
+    }
+
+    private fun applyLink(
+        url: String,
+        start: Int,
+        end: Int,
+        text: Spannable,
+    ) {
+        val urlSpanFactory = DEFAULT_SPAN_FACTORY
+        val span = urlSpanFactory(url)
+        text.setSpan(span, start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+    }
+
+    private fun makeUrl(
+        url: String,
+        prefixes: Array<String>,
+        matcher: Matcher,
+        filter: Linkify.TransformFilter?,
+    ): String {
+        var url = url
+        if (filter != null) {
+            url = filter.transformUrl(matcher, url)
+        }
+        var hasPrefix = false
+        for (i in prefixes.indices) {
+            if (url.regionMatches(0, prefixes[i], 0, prefixes[i].length, ignoreCase = true)) {
+                hasPrefix = true
+
+                // Fix capitalization if necessary
+                if (!url.regionMatches(0, prefixes[i], 0, prefixes[i].length, ignoreCase = false)) {
+                    url = prefixes[i] + url.substring(prefixes[i].length)
+                }
+                break
+            }
+        }
+        if (!hasPrefix && prefixes.isNotEmpty()) {
+            url = prefixes[0] + url
+        }
+        return url
+    }
+
+    private fun gatherLinks(
```

Comment:
Avoid to create functions that have `Unit` as the return type and apply a side effect to a parameter (`links` in tis case) and prefer to return the result of function. This makes it more functional and easier to reason about. 

You can just return the a `List` in this case.

Reply:
It should be considered more as effect than side effect. The `gatherLinks` responsibility is to search the link in the string and add it to the list.  It is the only effect of this method.